### DISCUSSION TITLE
How to store ID token when it's too big to add to Session?

### DISCUSSION THREAD

**Discussion by revmischa** [nextauthjs/next-auth#5732] [Help]
I'm using AWS Cognito and need to store the ID token alongside the access and refresh tokens. My problem is that if I store all of these in the Session object then the session cookie header becomes too large and get max header size violation errors. What's the best practice for storing these tokens? 

I should note that I want the idToken to be available to the client to send to a backend (lambda@edge) that is not nextjs and can only validate a cognito id token JWT, but doesn't have database access and doesn't know anything about next-auth. So I need the client to be able to read the id token.

My config:
```ts
import { decodeJwt } from 'jose';
import NextAuth, { NextAuthOptions, Session } from 'next-auth';
import CognitoProvider from 'next-auth/providers/cognito';
import { refreshTokensIfNeeded } from 'web/lib/client/auth';
import { COGNITO_CLIENT_ID, COGNITO_USER_POOL_ID, REGION } from 'web/lib/config/next';

if (!REGION) throw new Error('REGION is not set');
if (!COGNITO_CLIENT_ID) throw new Error('COGNITO_CLIENT_ID is not set');
if (!COGNITO_USER_POOL_ID) throw new Error('COGNITO_USER_POOL_ID is not set');

export const authOptions: NextAuthOptions = {
  providers: [
    CognitoProvider({
      clientId: COGNITO_CLIENT_ID,
      issuer: `https://cognito-idp.${REGION}.amazonaws.com/${COGNITO_USER_POOL_ID}`,

      // use cognito for token signing
      // https://github.com/nextauthjs/next-auth/issues/4707
      clientSecret: '',
      client: {
        token_endpoint_auth_method: 'none',
      },
      checks: 'nonce', // https://github.com/nextauthjs/next-auth/discussions/3551
      authorization: { params: { identity_provider: 'linkedin' } }, // skip cognito hosted UI
    }),
  ],
  // debug: true,
  pages: {
    signIn: '/login',
  },
  callbacks: {
    // need to handle token refresh https://next-auth.js.org/tutorials/refresh-token-rotation
    // return value is added to Session
    jwt: async ({ token, account, user }) => {
      // Initial sign in
      if (account && user) {
        const { id_token, access_token, refresh_token, expires_at } = account;
        const idTokenDecoded = decodeJwt(id_token as string);

        return {
          // save token to session for authenticating to AWS
          // https://next-auth.js.org/configuration/callbacks#jwt-callback
          accessToken: access_token,
          accessTokenExpires: expires_at ? expires_at * 1000 : 0,
          refreshToken: refresh_token,
          idToken: id_token,
          user,
        };
      }

      // already logged-in
      // refresh access token if needed
      const session: Session = token as Session;
      return await refreshTokensIfNeeded(session);
    },

    session: async ({ session, token }) => {
      if (!session?.user || !token?.accessToken) {
        console.error('No accessToken found on token or session');
        return session;
      }
      session.accessToken = token.accessToken as string;
      session.idToken = token.idToken as string;
      session.user = token.user as Session['user'];
      session.error = token.error as string | undefined;
      return session;
    },
  },
};

export default NextAuth(authOptions);
```

When I try storing the idToken in the session cookie in this manner I get this error from AWS (S3/nextjs edge):
```
<Code>RequestHeaderSectionTooLarge</Code>
<Message>Your request header section exceeds the maximum allowed size.</Message>
```
Even though the session cookie appears to be chunked, the cookie header itself is too large for AWS:
![image](https://user-images.githubusercontent.com/245131/200098488-705fbf3a-a937-40ba-a6ae-95abd919ca9b.png)

If i understand what is happening correctly, mixpanel cookies + next-auth-session-encrypted(cognito access+refresh+id tokens) > 8192kb of cookies which means the web browser client will never be able to access your website again because the cookie size will be too large.

This problem is coming up because I'm trying to persist the id token from cognito, which is rather large, in the nextjs auth session, which is stored in cookies. i'm trying to do this because i want to send the id token to an edge function which can validate custom claims about my token (tenant id).

<img width="1398" alt="Screen Shot 2022-11-04 at 8 53 07 PM" src="https://user-images.githubusercontent.com/245131/200099684-2ac4c9d6-091f-47a8-a4c4-4540a3f71f1a.png">

**Comment by clipartinc** [comment]
I had a similar error with mySQL where the id_token was too long. I had to edit the _/node_modules/@next_auth/typeorm-legacy-adapter/dist/entities.js_ file, I changed this...
```
__decorate([
    (0, typeorm_1.Column)({ type: "varchar", nullable: true }),
    __metadata("design:type", Object)
], AccountEntity.prototype, "id_token", void 0);
```

to...

```
__decorate([
    (0, typeorm_1.Column)({ type: "mediumtext", nullable: true }),
    __metadata("design:type", Object)
], AccountEntity.prototype, "id_token", void 0);
```

I had to then...
- remove all data from the accounts, sessions, and users columns to start fresh
- restart the app

You should now be able to save long sessions into id_token column. To make sure check the mySQL structure of id_token to make sure it's mediumtext after the restart.

Hope this can help some mySQL users.
